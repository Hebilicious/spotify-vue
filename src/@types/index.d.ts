export type Track = {
    _id: string,
    name: string,
    year: number | null,
    artist: string | null,
    album: string | null,
    duration_s: number,
    imageUri: string,
    type: string,
    uri: string
};

export type Album = {
    type: string,
    _id: string,
    name: string,
    year: number,
    total_tracks: number,
    description: string,
    imageUri: string,
    tracks: Track[]
};

export type PlayList = Album[]