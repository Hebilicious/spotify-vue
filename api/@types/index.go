package api

type Track struct {
	ID       string `json:"_id"`
	Name     string `json:"name"`
	Year     int    `json:"year"`
	Artist   string `json:"artist"`
	Album    string `json:"album"`
	Duration int    `json:"duration_s"`
	ImageUri string `json:"imageUri"`
	Type     string `json:"type"`
	Uri      string `json:"uri"`
}

type Album struct {
	ID          string  `json:"_id"`
	Type        string  `json:"type"`
	Name        string  `json:"name"`
	Year        int     `json:"year"`
	TotalTracks int     `json:"total_tracks"`
	Description string  `json:"description"`
	ImageUri    string  `json:"imageUri"`
	Tracks      []Track `json:"tracks"`
}

type PlayList []Album
