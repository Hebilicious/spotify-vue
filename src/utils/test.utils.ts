import { VueWrapper } from '@vue/test-utils';
import { expect, it } from 'vitest';

export const testSnapShot = (wrapperGetter: () => VueWrapper) => {
  it('Matches the snapshot', () => {
    expect(wrapperGetter().html()).toMatchSnapshot();
  });
};
