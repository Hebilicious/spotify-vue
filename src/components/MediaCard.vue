<template>
  <div class="media-card">
    <img :src="album.imageUri" alt="cover-art" />
    <div class="play-icon" @click="playCurrentAlbum">
      <div class="play-icon-circle">
        <span class="material-icons">play_arrow</span>
      </div>
    </div>
    <div class="media-title">{{ album.name }}</div>
    <span class="media-subtitle">{{ album.description }}</span>
  </div>
</template>

<script lang="ts" setup>
import { Album } from '@/@types';
import { useNowPlayingStore } from '@/stores/nowPlaying';

const nowPlayingStore = useNowPlayingStore();
const { album } = defineProps<{ album: Album }>();

const playCurrentAlbum = () => {
  nowPlayingStore.setCurrentPlaylist(album.tracks)
}
</script>

<style lang="scss">
.media-card {
  background-color: var(--color-background-soft);
  height: 250px;
  width: 185px;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  padding: 1rem;
  align-items: flex-start;
  transition: all 0.5s ease;

  img {
    height: 150px;
    width: 100%;
    border-radius: 5px;
    object-fit: cover;
  }

  .media-title {
    padding: 1rem 0.5rem 0.25rem 0;
    white-space: nowrap;
    width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: 16px;
    font-weight: 700;
    letter-spacing: normal;
    line-height: 24px;
    text-transform: none;
    color: var(--color-heading);
  }

  .play-icon {
    visibility: hidden;
    transform: translate(110px, 110px);
    position: absolute;
    opacity: 0;
    cursor: pointer;
    transition: visibility 0.5s ease, opacity 0.5s ease-in-out;

    .play-icon-circle {
      border-radius: 50%;
      background: #1fdf64;
      height: 50px;
      width: 50px;
      box-shadow: 0 8px 8px rgba(0, 0, 0, 0.3);

      transform: translate(-10px, -15px);
      span {
        transition: transform 0.5s ease-out;
        color: black;
        transform: translate(13px, 13px) scale(1);
      }
    }
  }

  &:hover {
    background-color: #252424;
    transform: scale(1.05);
    .play-icon {
      visibility: visible;
      opacity: 1;
      span {
        transform: translate(13px, 13px) scale(2);
      }
    }
  }
}
</style>
