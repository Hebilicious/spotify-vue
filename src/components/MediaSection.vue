<template>
  <div class="media-section">
    <div class="media-section-header">
      <div class="section-heading">{{ sectionTitle }}</div>
    </div>
    <div class="media-card-row">
      <template v-for="album in albums" :key="album._id">
        <MediaCard :album="album" />
      </template>
    </div>
  </div>
</template>

<script lang="ts" setup>
import MediaCard from './MediaCard.vue';

import { Album } from '@/@types';

export interface MediaSectionProps {
  sectionTitle: string;
  albums: Album[];
}

defineProps<MediaSectionProps>();
</script>

<style lang="scss">
.media-section {
  width: 100%;
  .media-section-header {
    display: flex;
    .section-heading {
      color: var(--color-heading);
      font-weight: 700;
      font-size: 1.5rem;
      flex: 1;
    }
  }
  .media-card-row {
    display: flex;
    width: 100%;
    padding-top: 1.75rem;
    grid-column-gap: 2rem;
  }
}
</style>
