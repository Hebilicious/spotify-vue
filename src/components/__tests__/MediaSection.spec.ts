import { VueWrapper, shallowMount } from '@vue/test-utils';
import { afterEach, beforeEach, describe, expect, fn, it } from 'vitest';
import { createTestingPinia } from '@pinia/testing';

import MediaCard from '@/components/MediaCard.vue';
import MediaSection, { MediaSectionProps } from '@/components/MediaSection.vue';

import { testSnapShot } from '@/utils/test.utils';
import { Album } from '@/@types';

describe('MediaSection', () => {
  let wrapper: VueWrapper;

  const mockMediaCards: Partial<Album>[] = [
    {
      imageUri: 'mock-url',
      name: 'mock-title',
      description: 'mock-subtitle',
    },
    {
      imageUri: 'mock-url',
      name: 'mock-title',
      description: 'mock-subtitle',
    },
    {
      imageUri: 'mock-url',
      name: 'mock-title',
      description: 'mock-subtitle',
    },
  ];

  const mockMediaSectionProps: MediaSectionProps = {
    sectionTitle: 'mock-section-1',
    albums: (mockMediaCards as Album[]),
  };

  beforeEach(() => {
    wrapper = shallowMount(MediaSection, {
      props: mockMediaSectionProps
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('displays the title of the section', () => {
    expect(wrapper.find('.section-heading').text()).toBe('mock-section-1');
  });

  it('renders media cards for each card passed', () => {
    expect(wrapper.findAllComponents(MediaCard).length).toBe(3);
  });

  testSnapShot(() => wrapper);
});
