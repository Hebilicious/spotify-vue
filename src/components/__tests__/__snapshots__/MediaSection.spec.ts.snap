// Vitest Snapshot v1

exports[`MediaSection > Matches the snapshot 1`] = `
"<div class=\\"media-section\\">
  <div class=\\"media-section-header\\">
    <div class=\\"section-heading\\">mock-section-1</div>
  </div>
  <div class=\\"media-card-row\\">
    <media-card-stub album=\\"[object Object]\\"></media-card-stub>
    <media-card-stub album=\\"[object Object]\\"></media-card-stub>
    <media-card-stub album=\\"[object Object]\\"></media-card-stub>
  </div>
</div>"
`;
